Merge feature/capgen into main 2021/08/12

climbfuji opened the pull request:
Update main with the latest development in `feature/capgen`.

Where incompatibilities exist between `capgen` and `ccpp_prebuild`, a workaround was added on the `ccpp_prebuild` side. Exception: in `parse_checkers.py`, a temporary workaround is added to internally convert unit `none` to `1` to match the ccpp-physics/UFS/SCM units for dimensionless variables (`none`) with what `capgen` uses (`1`). A discussion has been started to eliminate those inconsistencies in the actual metadata (see https://github.com/ESCOMP/CCPPStandardNames/discussions/18).

Update 08/18/2021: one additional routine had to be added to `parse_checkers.py` that I forgot to mention previously: `registered_fortran_ddt_names` - this routine is required so that `ccpp_prebuild.py` can get access to all the registered Fortran DDTs, because it utilizes capgen's metadata parser on a file-by-file basis.

Associated PRs:

https://github.com/NCAR/ccpp-framework/pull/391
https://github.com/earth-system-radiation/rte-rrtmgp/pull/133
https://github.com/NCAR/ccpp-physics/pull/718
https://github.com/NOAA-EMC/fv3atm/pull/367
https://github.com/ufs-community/ufs-weather-model/pull/745

For regression testing, see https://github.com/ufs-community/ufs-weather-model/pull/745.

climbfuji commented:
All, we are planning to merge this PR next week. Please review when you have time. It is important to bring the latest updates to the capgen metadata parser into the UFS.

Once the PR is merged, I will create a PR back from main to feature/capgen to synchronize the two branches. This will be much easier to see the diffs, and if there are any changes that are incompatible with capgen (I am not aware of any), then we can correct them at that point.

Thanks!

gold2718 commented:
> All, we are planning to merge this PR next week. Please review when you have time. It is important to bring the latest updates to the capgen metadata parser into the UFS.

To be clear, are you asking me to prioritize this over any work on capgen?

climbfuji commented:
> > All, we are planning to merge this PR next week. Please review when you have time. It is important to bring the latest updates to the capgen metadata parser into the UFS.
> 
> To be clear, are you asking me to prioritize this over any work on capgen?

No. As usual, reviewers who do not have the time to review the PRs can take themselves off the list. By providing everyone with a timeline, reviewers can judge better whether they will have the time or not. What goes into master right now isn't that important for you, more important will be the reverse PR - merging main back to feature/capgen. For that one, the changes will be much smaller, especially for the capgen related files.

gold2718 commented:
Right now, I do not understand the **need** for a reverse merge into capgen (which then requires me to merge to all my branches). To understand that need, I need to understand this PR. In my opinion, it's pay now or pay later.

climbfuji commented:
Synchronizing them now means fewer merge conflicts, fewer functional differences and surprises down the road. It's helpful in my opinion, but if it is considered a burden then we don't need to do the reverse PR.

> On Aug 13, 2021, at 2:50 PM, goldy ***@***.***> wrote:
> 
> 
> Right now, I do not understand the need for a reverse merge into capgen (which then requires me to merge to all my branches). To understand that need, I need to understand this PR. In my opinion, it's pay now or pay later.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/NCAR/ccpp-framework/pull/391#issuecomment-898713026>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5C2RIKUCEI5RFVH4HQLETT4WARDANCNFSM5CBPBDVA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
>

grantfirl reviewed (approved):
A careful review of this PR is probably beyond my expertise right now since I'm not very familiar with capgen. To the extent that ccpp_prebuild.py calls capgen's parsers, as long as it continues to work, I don't have a problem approving the PR though. Looking at the associated fv3atm PR, it doesn't look there are a bunch of changes to understand there, perhaps with the exception of adding ccpp_types.F90 to VARIABLE_DEFINITION_FILES in the config file. I'm assuming this is now necessary due to updates in the Fortran parser or something? Also, if the order of files in that list should have DDT definitions first, why did it not error out before this PR? Is this a new requirement (I didn't think so -- I thought this is mentioned in the docs)?

llpcarson reviewed (approved):
Merge looks good, approved.

climbfuji commented:
> A careful review of this PR is probably beyond my expertise right now since I'm not very familiar with capgen. To the extent that ccpp_prebuild.py calls capgen's parsers, as long as it continues to work, I don't have a problem approving the PR though. Looking at the associated fv3atm PR, it doesn't look there are a bunch of changes to understand there, perhaps with the exception of adding ccpp_types.F90 to VARIABLE_DEFINITION_FILES in the config file. I'm assuming this is now necessary due to updates in the Fortran parser or something? Also, if the order of files in that list should have DDT definitions first, why did it not error out before this PR? Is this a new requirement (I didn't think so -- I thought this is mentioned in the docs)?

Thanks for these comments. A few answers to your questions:
1. The order of DDTs didn't matter in the past, because we were collecting data from the `capgen.py` metadata parser, which didn't check if a DDT was defined when encountering a reference to it. The consistency check happened later in `ccpp_prebuild.py`, after all data was collected and merged. Now, `capgen.py`'s metadata parser does check if the DDTs are known and therefore enforces the requirement of having the correct order of the variable definition files.
2. `ccpp_prebuild.py` had the `ccpp_t` type (in `ccpp_types.F90`, `ccpp_types.meta`) hardcoded into the scripts instead of reading it from the files. Because of 1., this data must now be read by the `capgen.py` metadata parser.

In summary, these PRs correct several inconsistencies i the metadata w.r.t. derived data types that `ccpp_prebuild.py` simply didn't care about, but `capgen.py` does (for good reasons). This is not only a good thing to do, but also necessary for switching to `capgen.py` later.

gold2718 commented on scripts/metadata_table.py:
```diff
@@ -96,475 +125,1137 @@
+                        # end if
+                        self.__table_type = value
+                    elif key == 'dependencies':
+                        if value.lower() == "none" or value.strip() == '':
```
Why is a blank `dependencies` line an error here?

climbfuji replied:
capgen's metadata parser was creating a dependencies list that contained an item `[ ' ' ]` - this produced an error.

gold2718 replied:
That will not be happening once you merge #395 (and should not have been happening with the change you made).

gold2718 commented on scripts/metadata_table.py:
```diff
@@ -96,475 +125,1137 @@
+                            emsg += "already been declared as None"
+                            self.__pobj.add_syntax_err(emsg)
+                        else:
+                            depends = [x.strip() for x in value.split(',') if x.strip()]
```
This change should be made in feature/capgen **first**, then merged into prebuild. Doing it this way will cause issues later on.

climbfuji replied:
If there is a PR to feature/capgen that I can pull in that is identical to this change - and nothing else - then I can pull it in, fix the merge conflict, assert that nothing has changed in the codebase, and don't have to rerun the tests. I don't see however why this change is causing a problem. It can be merged back from main to feature/capgen without issues.

gold2718 replied:
See #395

gold2718 replied:
> I don't see however why this change is causing a problem. 

For one thing, it took me 20 times longer to process this PR due to this change and sloppy merging.

gold2718 commented on scripts/metadata_table.py:
```diff
@@ -96,475 +125,1137 @@
+        # end if
+        if self.header_type == "ddt":
+            known_ddts.append(self.title)
+        # end if
```
There should be code here which is missing. What happened?

climbfuji replied:
I am not sure, there were several merge conflicts that I had to resolve. What is missing here?

gold2718 replied:
Do a diff against the feature/capgen version. After you merge #395, there should be no difference.
For this PR to be okay, there needs to be only one change between capgen code in feature/capgen and the code in this PR. Currently, that is not true.

gold2718 commented on scripts/metavar.py:
```diff
@@ -183,111 +266,147 @@ class VariableProperty(object):
+    >>> VariableProperty('local_name', str, check_fn_in=check_local_name).valid_value('foo(bar)')
+    'foo(bar)'
+    >>> VariableProperty('local_name', str, check_fn_in=check_local_name).valid_value('q(:,:,index_of_water_vapor_specific_humidity)')
+    'q(:,:,index_of_water_vapor_specific_humidity)'
```
The variables below have been removed in feature/capgen. Why are they back?

climbfuji replied:
git seems to have added this code back in apparently when I did the merge

gold2718 reviewed (changes requested):
- I do not understand the changes to parse_checkers.py. The command:
```
git diff NCAR/feature/capgen dom/merge_feature_capgen_into_main_20210812 -- scripts/parse_tools/parse_checkers.py
```
shows more changes than the unit conversion from `none` to `1`, however, these changes are not showing up in the diffs in this PR. Please explain.
Note: NCAR stands for github.com:/NCAR and dom stands for github.com/climbfuji (git command assumes that both remotes are loaded in a sandbox).
- convert_metadata.py is in feature/capgen but not in this PR. Did you decide to remove it?
- metavar.py does not match feature/capgen. Why?

climbfuji commented on .travis.yml:
```diff
@@ -0,0 +1,28 @@
+language: python
+
+python:
+  - "2.7"
```
We don't use Python 2 anymore. I think this can be upgraded safely to Python 3 in a follow-up PR.

climbfuji commented:
> * I do not understand the changes to parse_checkers.py. The command:
> 
> ```
> git diff NCAR/feature/capgen dom/merge_feature_capgen_into_main_20210812 -- scripts/parse_tools/parse_checkers.py
> ```
> 
> shows more changes than the unit conversion from `none` to `1`, however, these changes are not showing up in the diffs in this PR. Please explain.
> Note: NCAR stands for github.com:/NCAR and dom stands for github.com/climbfuji (git command assumes that both remotes are loaded in a sandbox).
> 
> * convert_metadata.py is in feature/capgen but not in this PR. Did you decide to remove it?
> * metavar.py does not match feature/capgen. Why?

* `convert_metadata.py` is no longer needed, there is no model that uses the old metadata standard
* I will need to look at the diffs of `metavar.py` when I get back to my work laptop to see why they don't match

climbfuji commented:
@gold2718 Thank you for your thorough code review. I would appreciate if you could keep these conversation at a professional level and refrain from comments like https://github.com/NCAR/ccpp-framework/pull/391#discussion_r690888730.

Several of the changes that you complained about were made in main (or master at that time) **in agreement and with your approval**. They are **not** the result of a sloppy merge.

1. Why is `metavar.py` different? Because `feature/capgen` still does not support the active attribute. In PR #307, the capability to process the `active` attribute was added to main's `metavar.py`, and it was review and approved. This needs to stay to keep the system working.
2. Similar, in `metavar.py`, the update to `true` and `false` attributes were made in PR https://github.com/NCAR/ccpp-framework/pull/205, which was reviewed and approved. This needs to stay to keep the system working (ccpp_prebuild uses it).
3. metadata_table.py: the changes to the documentation (example metadata table, lines 61-96) were made in PRs https://github.com/NCAR/ccpp-framework/pull/317 and https://github.com/NCAR/ccpp-framework/pull/205, which were reviewed and approved. They are still valid and should stay.

I made the following corrections based on your code review in commit https://github.com/NCAR/ccpp-framework/pull/391/commits/53aa9dc7cffd8fd0c7dce57902b8d5b1345d3fdb:

4. metadata_table.py: The change in line 377 is no longer required. I think this was needed to get past a parsing issue before line 392 was fixed in my PR (and then in your PR #395, which I merged into my PR to clean up the commit history).
5. metadata_table.py: The additional lines 736-741 are not required. They were a leftover of the original "dependencies" implementation in the main branch, which is replaced with the feature/capgen implementation as part of this PR. I removed them
6. metadata_table.py: I am not sure why the lines `if self.module is None` didn't get merged into my branch from feature/capgen, but I added them back (now lines 778-781).
7. parse_checkers.py: I forgot to mention in the PR description that I needed to add one additional routine to parse_checkers.py: registered_fortran_ddt_names. This routine is required so that ccpp_prebuild.py can get access to all the registered Fortran DDTs, because it utilizes capgen's metadata parser on a file-by-file basis. I updated the PR description accordingly.
8. parse_checkers.py: reinstantiated the missing comment line 10 "# CCPP framework imports" and removed LITERAL_INT (line 206-ish).

The fact that we are seeing more changes that slowed down your code review is not only because of the oversights listed in 4-8, but also because feature/capgen has not been updated from main (master) for years. I am unsure why this didn't happen. In all other projects I have been working, the maintainers of feature branches are usually keeping their branches up to date with the authoritative code to avoid exactly these kind of problems, especially when they are long-term developments like feature/capgen. But maybe there was a misunderstanding and we (DTC) should have created those PRs instead of you. You would still have had to review and accept them, though.

After fixing the few merge problems described above (items 4, 5, 6, 8) and updating the PR description (item 7), the current PR must be merged as-is to keep the system capgen-metadata-parser + ccpp_prebuild working. We do not have the time and I do not see why I should be making the extra efforts to make these changes separately in feature/capgen first at this point, then merge them into my PR and fix the merge conflicts. It is natural in distributed development that changes are made in the same files in different branches, but as I said above they should have been merged back to feature/capgen immediately afterwards instead of years later.

To avoid such problems in the future, I will create a follow-up PR to bring the changes/differences described above back to feature/capgen, and I strongly encourage that we do so on a regular basis going forward.

Fortunately, after making all of the abovee changes, the auto-generated code is 100% identical to before, which means we do not have to rerun the regression tests on all nine platforms (we will rerun them on just one with two different compilers).

gold2718 commented:
> Why is metavar.py different? Because feature/capgen still does not support the active attribute. In PR Update master from gsd/develop 2020/06/30 #307, the capability to process the active attribute was added to main's metavar.py, and it was review and approved. This needs to stay to keep the system working.

Then, as part of the capgen unification project, this change should be made in the feature/capgen branch. Continually updating the main branch without contributing to the feature/capgen branch is slowing down the unification effort.

gold2718 commented:
> Similar, in metavar.py, the update to true and false attributes were made in PR Migrate to new metadata format step 1 #205, which was reviewed and approved. This needs to stay to keep the system working (ccpp_prebuild uses it).

If prebuild is using a feature not used by feature/capgen, that code should be moved to the prebuild codebase.

gold2718 commented:
> metadata_table.py: the changes to the documentation (example metadata table, lines 61-96) were made in PRs Add dependencies to metadata #317 and Migrate to new metadata format step 1 #205, which were reviewed and approved. They are still valid and should stay.

Please move these changes to feature/capgen, then merge from there.

climbfuji commented:
> > Similar, in metavar.py, the update to true and false attributes were made in PR Migrate to new metadata format step 1 #205, which was reviewed and approved. This needs to stay to keep the system working (ccpp_prebuild uses it).
> 
> If prebuild is using a feature not used by feature/capgen, that code should be moved to the prebuild codebase.

No. We discussed this in the ccpp-framework meeting at that time, it was decided to generalize the false/true logic and I followed your recommendations at that time to do. This is an improvement for capgen that we agreed upon, which prebuild has adopted since then. I am not going to revert or move it.

gold2718 commented:
> The fact that we are seeing more changes that slowed down your code review is not only because of the oversights listed in 4-8, but also because feature/capgen has not been updated from main (master) for years. I am unsure why this didn't happen. In all other projects I have been working, the maintainers of feature branches are usually keeping their branches up to date with the authoritative code to avoid exactly these kind of problems, especially when they are long-term developments like feature/capgen. 

I agree, when are you going to do this? I argue that these changes should have been made first in feature/capgen in order to make progress towards capgen unification.

climbfuji commented:
> > Why is metavar.py different? Because feature/capgen still does not support the active attribute. In PR Update master from gsd/develop 2020/06/30 #307, the capability to process the active attribute was added to main's metavar.py, and it was review and approved. This needs to stay to keep the system working.
> 
> Then, as part of the capgen unification project, this change should be made in the feature/capgen branch. Continually updating the main branch without contributing to the feature/capgen branch is slowing down the unification effort.

I will create a PR with the active attribute changes, the true/false (valid vals) changes, and the documentation changes for feature/capgen, merge it, then formally update this PR (which must result in no changes to the code), then this should be ok.

climbfuji commented:
> > > Why is metavar.py different? Because feature/capgen still does not support the active attribute. In PR Update master from gsd/develop 2020/06/30 #307, the capability to process the active attribute was added to main's metavar.py, and it was review and approved. This needs to stay to keep the system working.
> > 
> > 
> > Then, as part of the capgen unification project, this change should be made in the feature/capgen branch. Continually updating the main branch without contributing to the feature/capgen branch is slowing down the unification effort.
> 
> I will create a PR with the active attribute changes, the true/false (valid vals) changes, and the documentation changes for feature/capgen, merge it, then formally update this PR (which must result in no changes to the code), then this should be ok.

The PR is here: https://github.com/NCAR/ccpp-framework/pull/396 - please merge and I will pull it into this PR.

climbfuji commented:
> > The fact that we are seeing more changes that slowed down your code review is not only because of the oversights listed in 4-8, but also because feature/capgen has not been updated from main (master) for years. I am unsure why this didn't happen. In all other projects I have been working, the maintainers of feature branches are usually keeping their branches up to date with the authoritative code to avoid exactly these kind of problems, especially when they are long-term developments like feature/capgen.
> 
> I agree, when are you going to do this? I argue that these changes should have been made first in feature/capgen in order to make progress towards capgen unification.

Immediately after https://github.com/NCAR/ccpp-framework/pull/391 is merged.

climbfuji commented:
@gold2718 after merging https://github.com/NCAR/ccpp-framework/pull/396 into feature/capgen and merging feature/capgen into this branch again, the files `metadata_table.py` and `metavar.py` are identical between the head of feature/capgen and this branch/PR. I am going tor re-request your review now.

gold2718 reviewed (approved):
Looks good now, thanks!
